### DISCUSSION TITLE
Recomended way to manage users?

### DISCUSSION THREAD

**Discussion by jorgerpo** [fastapi/fastapi#8232] [Questions]
**Description**

What is the best way to handle users registers, login, pasword changes, unregister?
A way to manage permissions and roles?
**Additional context**
Something similar to flask-auth , flask -security, flask login?

**Comment by tiangolo** [comment]
Sorry for the delay, it would probably be with Dependencies: https://fastapi.tiangolo.com/tutorial/dependencies/first-steps/

As an example, check the project generators:

https://github.com/tiangolo/full-stack-fastapi-postgresql

https://github.com/tiangolo/full-stack-fastapi-couchbase

**Comment by tiangolo** [comment]
I understand these links answer the question, so I'll close this issue. But feel free to add more comments or create new issues.

**Comment by frankie567** [comment]
Just to put my two cents in, I'm currently working on a library proposing ready-to-use auth-related routes and base user model: https://github.com/frankie567/fastapi-users

The goal is to deliver the basic/boring/ever-repeating logic while giving the developer enough room to customize it to their needs.

**Comment by euri10** [comment]
excellent !

On Fri, Oct 25, 2019 at 9:04 AM François Voron <notifications@github.com>
wrote:

> Just to put my two cents in, I'm currently working on a library proposing
> ready-to-use auth-related routes and base user model:
> https://github.com/frankie567/fastapi-users
>
> The goal is to deliver the basic/boring/ever-repeating logic while giving
> the developer enough room to customize it to their needs.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/issues/67?email_source=notifications&email_token=AAINSPTOJBHFNDZCMU6W4QDQQKK63A5CNFSM4G3W3URKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECHM6GA#issuecomment-546230040>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINSPVFQRK3XVECOKYQFC3QQKK63ANCNFSM4G3W3URA>
> .
>


-- 
benoit barthelet
http://pgp.mit.edu/pks/lookup?op=get&search=0xF150E01A72F6D2EE

**Comment by Raphasha27** [comment]
Great question. I built a similar auth system for a fintech backend - here's the recommended approach for FastAPI.

## Core Stack

- **FastAPI** with `Depends()` for DI
- **SQLAlchemy** for user model/persistence
- **passlib[bcrypt]** for password hashing
- **python-jose** for JWT tokens
- **OAuth2PasswordBearer** for token extraction

## User Model

```python
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Enum
from sqlalchemy.sql import func
from passlib.context import CryptContext
import enum

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

class UserRole(str, enum.Enum):
    ADMIN = "admin"
    USER = "user"
    MODERATOR = "moderator"

class User(Base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True)
    email = Column(String, unique=True, index=True, nullable=False)
    hashed_password = Column(String, nullable=False)
    role = Column(Enum(UserRole), default=UserRole.USER)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    def set_password(self, password: str):
        self.hashed_password = pwd_context.hash(password)

    def verify_password(self, password: str) -> bool:
        return pwd_context.verify(password, self.hashed_password)
```

## Registration & Login

```python
from fastapi import FastAPI, Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from jose import JWTError, jwt
from datetime import datetime, timedelta

app = FastAPI()
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/auth/login")

SECRET_KEY = "your-secret-key"  # use env var
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 60

@app.post("/auth/register")
async def register(email: str, password: str, db: Session = Depends(get_db)):
    existing = db.query(User).filter(User.email == email).first()
    if existing:
        raise HTTPException(status_code=400, detail="Email already registered")
    user = User(email=email)
    user.set_password(password)
    db.add(user)
    db.commit()
    db.refresh(user)
    return {"id": user.id, "email": user.email}

@app.post("/auth/login")
async def login(form: OAuth2PasswordRequestForm = Depends(), db: Session = Depends(get_db)):
    user = db.query(User).filter(User.email == form.username).first()
    if not user or not user.verify_password(form.password):
        raise HTTPException(status_code=401, detail="Invalid credentials")
    token = jwt.encode(
        {"sub": str(user.id), "role": user.role.value, "exp": datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)},
        SECRET_KEY, algorithm=ALGORITHM
    )
    return {"access_token": token, "token_type": "bearer"}
```

## Role-Based Access Control

```python
from functools import wraps

def require_role(required_role: UserRole):
    def dependency(current_user: User = Depends(get_current_user)):
        if current_user.role != required_role and current_user.role != UserRole.ADMIN:
            raise HTTPException(status_code=403, detail="Insufficient permissions")
        return current_user
    return dependency

@app.get("/admin/users")
async def list_users(admin: User = Depends(require_role(UserRole.ADMIN)), db: Session = Depends(get_db)):
    return db.query(User).all()
```

## Password Changes

```python
@app.post("/auth/change-password")
async def change_password(
    old_password: str, new_password: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    if not current_user.verify_password(old_password):
        raise HTTPException(status_code=400, detail="Incorrect password")
    current_user.set_password(new_password)
    db.commit()
    return {"message": "Password updated"}
```

## Key Differences from Flask-Login

| Flask-Login | FastAPI Equivalent |
|-------------|-------------------|
| `login_required` | `Depends(get_current_user)` |
| `current_user` proxy | `Depends()` injection |
| Session-based | JWT/token-based |
| `@login_manager.user_loader` | `get_current_user` dependency |

FastAPI's dependency injection system makes this cleaner than Flask's decorator approach - each route explicitly declares what auth it needs, making the security model visible in the function signature.
